internal/core/cppapi_data.cpp

```diff
@@ -327,55 +257,50 @@ void horizon(
     auto xline  = metadata.xline();
     auto sample = metadata.sample();
 
-    std::size_t const nsamples = buffer_offsets[to] - buffer_offsets[from];
+    std::size_t const nsamples = subvolume.nsamples(from, to);
     if (nsamples == 0){
         return;
     }
     std::unique_ptr< voxel[] > samples(new voxel[nsamples]{{0}});
 
-    VerticalWindow window(sample.stepsize(), 2, sample.min());
     std::size_t cur = 0;
     for (int i = from; i < to; ++i) {
-        if(buffer_offsets[i] == buffer_offsets[i+1]) {
+        if (subvolume.is_empty(i)) {
             continue;
         }
 
-        window.move(reference[i] - top[i], bottom[i] - reference[i]);
-
-        double nearest_reference_depth = window.nearest(reference[i]);
-
-        auto const cdp = horizontal.to_cdp(i);
-        auto ij = transform.WorldToAnnotation({cdp.x, cdp.y, 0});
+        auto segment = subvolume.vertical_segment(i);
 
-        double nearest_top_depth    = nearest_reference_depth -
-                                      window.nsamples_above() * window.stepsize();
-        double nearest_bottom_depth = nearest_reference_depth +
-                                      window.nsamples_below() * window.stepsize();
+        double top_sample_depth    = segment.top_sample_position();
+        double bottom_sample_depth = segment.bottom_sample_position();
 
-        if (not sample.inrange(nearest_top_depth) or
-            not sample.inrange(nearest_bottom_depth))
+        if (not sample.inrange(top_sample_depth) or
+            not sample.inrange(bottom_sample_depth))
         {
-            auto row = horizontal.row(i);
-            auto col = horizontal.col(i);
+            auto row = horizontal_grid.row(i);
+            auto col = horizontal_grid.col(i);
             throw std::runtime_error(
                 "Vertical window is out of vertical bounds at"
                 " row: " + std::to_string(row) +
                 " col:" + std::to_string(col) +
-                ". Request: [" + std::to_string(nearest_top_depth) +
-                ", " + std::to_string(nearest_bottom_depth) +
+                ". Request: [" + std::to_string(top_sample_depth) +
+                ", " + std::to_string(bottom_sample_depth) +
                 "]. Seismic bounds: [" + std::to_string(sample.min())
                 + ", " +std::to_string(sample.max()) + "]"
             );
         }
 
+        auto const cdp = horizontal_grid.to_cdp(i);
+        auto ij = transform.WorldToAnnotation({cdp.x, cdp.y, 0});
+
         ij[0]  = iline.to_sample_position(ij[0]);
         ij[1]  = xline.to_sample_position(ij[1]);
```

Comment:
We have an index in the user defined grid (i).
We translate it into world coordinates, as this is only general transformation available to us (cdp). World coordinates is external global coordinate system.
Now we can translate world data to the file coordinate system - annotation in our case (ij). Can't translate the depth axis because world data doesn't contain this information.
And finally you can translate to the system required by the openvds functions that read stuff by using to_sample_position which calls openvds function and requires annotation on input (hence why we translated into it and not to something else).

That seems like a big "what" comment to me that comes our from not yet being familiar with various coordinate systems available to us.

To me the only viable comment seems to be
"Translate user defined index -> world -> vds file annotation -> sample position needed to read data", but you figured out the same yourself already, so I am not sure it gives anything there.